tensorflow/compiler/tf2xla/kernels/image_projective_transform_op.cc

```diff
@@ -0,0 +1,326 @@
+  } else if (mode == Mode::FILL_CONSTANT) {
+    return in;
+  } else if (mode == Mode::FILL_NEAREST) {
+    return xla::Clamp(xla::ScalarLike(out, 0.0), out,
+                      xla::ScalarLike(out, length - 1));
+  }
+
+  return in;
+}
+
+class ImageProjectiveTransformOp : public XlaOpKernel {
+ public:
+  explicit ImageProjectiveTransformOp(OpKernelConstruction* context)
+      : XlaOpKernel(context) {
+    std::string interpolation_str;
+    OP_REQUIRES_OK(context,
+                   context->GetAttr("interpolation", &interpolation_str));
+    if (interpolation_str == "NEAREST") {
+      interpolation_ = Interpolation::INTERPOLATION_NEAREST;
+    } else if (interpolation_str == "BILINEAR") {
+      interpolation_ = Interpolation::INTERPOLATION_BILINEAR;
+    } else {
+      LOG(ERROR) << "Invalid interpolation " << interpolation_str
+                 << ". Supported types: NEAREST, BILINEAR";
+    }
+    std::string mode_str;
+    OP_REQUIRES_OK(context, context->GetAttr("fill_mode", &mode_str));
+    if (mode_str == "REFLECT") {
+      fill_mode_ = Mode::FILL_REFLECT;
+    } else if (mode_str == "WRAP") {
+      fill_mode_ = Mode::FILL_WRAP;
+    } else if (mode_str == "CONSTANT") {
+      fill_mode_ = Mode::FILL_CONSTANT;
+    } else if (mode_str == "NEAREST") {
+      fill_mode_ = Mode::FILL_NEAREST;
+    } else {
+      LOG(ERROR) << "Invalid mode " << mode_str
+                 << ". Supported types: REFLECT, WRAP, CONSTANT, NEAREST";
+    }
+  }
+
+  void Compile(XlaOpKernelContext* ctx) override {
+    const TensorShape& images_shape = ctx->InputShape(0);
+    const TensorShape& transforms_shape = ctx->InputShape(1);
+    OP_REQUIRES(ctx, images_shape.dims() == 4,
+                errors::InvalidArgument("Input images must have rank 4"));
+    OP_REQUIRES(ctx,
+                (TensorShapeUtils::IsMatrix(transforms_shape) &&
+                 (transforms_shape.dim_size(0) == images_shape.dim_size(0) ||
+                  transforms_shape.dim_size(0) == 1) &&
+                 transforms_shape.dim_size(1) == 8),
+                errors::InvalidArgument(
+                    "Input transform should be num_images x 8 or 1 x 8"));
+
+    TensorShape output_shape{images_shape.dim_size(1),
+                             images_shape.dim_size(2)};
+    if (ctx->num_inputs() >= kNumInputsV2) {
+      ctx->ConstantInputAsShape(2, &output_shape);
+      OP_REQUIRES(ctx, output_shape.dims() == 2,
+                  errors::InvalidArgument("output shape must be 1-dimensional",
+                                          output_shape.DebugString()));
+    }
+
+    OP_REQUIRES(ctx,
+                output_shape.dim_size(0) > 0 && output_shape.dim_size(1) > 0,
+                errors::InvalidArgument("output dimensions must be positive"));
+
+    output_shape.InsertDim(0, images_shape.dim_size(0));
+    output_shape.InsertDim(3, images_shape.dim_size(3));
+
+    xla::XlaOp fill_value;
+    if (ctx->num_inputs() >= kNumInputsV3) {
```

Comment:
Do you need this? There is no ImageProjectiveTransformV2 in xla op I think?

Reply:
I think it's even unnecessary to register V2 XLA op as XLA kernel appears after V3 commit?